[torch] Add cuda support for segment reduction 'max'

serhaty opened the pull request:
Summary:
This is re submit of PR: https://github.com/pytorch/pytorch/pull/54175

Main changes compared to original PR:
- Switch to importing "<ATen/cuda/cub.cuh>"
- Use CUB_WRAPPER to reduce boiler plate code.

Test Plan:
Will check CI status to make sure a

Added unit test

Differential Revision: D27941257

serhaty commented:
I am seeing two rocm related build issues but I am not sure what is the root cause since I don't see the output from UI. Is there a recommended way to debug such issues with rocm? Do I need to find an AMD capable device for this? cc: @jeffdaily for guidance on this.

ngimel commented:
@zasdfgbnm can you please add a note on the proper way to include new cub functions now? This PR is building ok with cuda, but breaks with ROCM.

zasdfgbnm commented:
I think for this PR, the problem is because `cub::DeviceSegmentedReduce` is not converted to `hipcub::DeviceSegmentedReduce` by hipify.

zasdfgbnm commented:
I think you need to add your functions to https://github.com/pytorch/pytorch/blob/master/torch/utils/hipify/cuda_to_hip_mappings.py

ngimel commented:
> I think you need to add your functions to https://github.com/pytorch/pytorch/blob/master/torch/utils/hipify/cuda_to_hip_mappings.py

And probably use NO_ROCM with CUB_WRAPPER

jeffdaily commented:
CC @jithunnair-amd @sunway513

serhaty commented:
> I think you need to add your functions to https://github.com/pytorch/pytorch/blob/master/torch/utils/hipify/cuda_to_hip_mappings.py

Just curious, would not it be better if default behavior was to do cub conversion automatically and only provide mapping if there is need for override (like API differences across the two)?

ngimel commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -8937,4 +8937,4 @@
 - func: segment_reduce(Tensor data, str reduce, *, Tensor? lengths=None, Tensor? indices=None, int axis=0, bool unsafe=False) -> Tensor
```
as we've discussed, add `initial` argument here to handle 0-length case

serhaty replied:
will leave this change to another PR on the stack in favor of smaller PRs.

ngimel reviewed (commented):
As we've discussed, please add `initial` argument to the handle

serhaty commented on aten/src/ATen/native/cuda/SegmentReduce.cu:
```diff
@@ -0,0 +1,93 @@
+struct CustomMax {
+  template <typename OutputT>
+  __host__ __device__ __forceinline__ OutputT
+  operator()(const OutputT& a, const OutputT& b) {
```
Thank you. Finally realized in logs that there is actually a way to see full logs  as well on jenkins.

Commit:
[torch] Add cuda support for segment reduction 'max' (#56704)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/56704

This is re submit of PR: https://github.com/pytorch/pytorch/pull/54175

Main changes compared to original PR:
- Switch to importing "<ATen/cuda/cub.cuh>"
- Use CUB_WRAPPER to reduce boiler plate code.

Test Plan:
Will check CI status to make sure a

Added unit test

Reviewed By: ngimel

Differential Revision: D27941257

fbshipit-source-id: 25d8e6fa7925ed9585c8c757e037fcb3d0237687
